compiler/noirc_evaluator/src/ssa/opt/array_get_from_if_else_result.rs

```diff
@@ -0,0 +1,248 @@
+        //
+        //     v13 = array_get v3, index v4
+        let else_result = dfg.insert_instruction_and_results(
+            Instruction::ArrayGet { array: else_value, index: *index },
+            block,
+            Some(element_types.clone()),
+            call_stack.clone(),
+        );
+        let else_result = else_result.first();
+
+        // Finally create an IfElse instruction like this:
+        //
+        //     v14 = if v0 then v12 else if v1 then v13
+        let new_result = dfg.insert_instruction_and_results(
+            Instruction::IfElse {
+                then_condition,
+                then_value: then_result,
+                else_condition,
+                else_value: else_result,
+            },
+            block,
+            None,
+            call_stack,
+        );
+        let new_result = new_result.first();
+
+        // And replace the original instruction's value with this final value
+        let results = dfg.instruction_results(instruction_id);
+        let result = results[0];
+        dfg.set_value_from_id(result, new_result);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::rc::Rc;
+
+    use crate::ssa::{
+        function_builder::FunctionBuilder,
+        ir::{
+            instruction::{Binary, Instruction},
+            map::Id,
+            types::Type,
+        },
+    };
+
+    #[test]
+    fn check_array_get_from_if_else_result_optimization() {
+        // acir(inline) fn main f0 {
+        //   b0(v0: [Field; 3], v1: [Field; 3], v2: u1, v3: u32):
+        //     v4 = not v2
+        //     v5 = if v2 then v0 else if v4 then v1
+        //     v6 = array_get v5, index v3
+        //     (no terminator instruction)
+        // }
+
+        let main_id = Id::test_new(0);
+        let mut builder = FunctionBuilder::new("main".into(), main_id);
+        let v0 = builder.add_parameter(Type::Array(Rc::new(vec![Type::field()]), 3));
+        let v1 = builder.add_parameter(Type::Array(Rc::new(vec![Type::field()]), 3));
+        let v2 = builder.add_parameter(Type::bool());
+        let v3 = builder.add_parameter(Type::unsigned(32));
+
+        let v4 = builder.insert_not(v2);
+        let v5 = builder
+            .insert_instruction(
+                Instruction::IfElse {
+                    then_condition: v2,
+                    then_value: v0,
+                    else_condition: v4,
+                    else_value: v1,
+                },
+                None,
+            )
+            .first();
+        builder.insert_array_get(v5, v3, Type::field());
+
+        let ssa = builder.finish();
+        println!("{ssa}");
+
+        // Expected output:
+        // acir(inline) fn main f0 {
+        //   b0(v0: [Field; 3], v1: [Field; 3], v2: u1, v3: u32):
+        //       v4 = not v2
+        //       v5 = if v2 then v0 else if v4 then v1
+        //       v7 = array_get v0, index v3
+        //       v8 = array_get v1, index v3
```

Comment:
I was expecting 
```
v7 = array_get v0, index v3
v8 = array_get v1, index v3
v9 = if v2 then v7 else if v4 then v8
v5 = if v2 then v0 else if v4 then v1
```

Reply:
We may have to change the order in which we find where we apply this optimization to create the SSA commented above.

Right now once we find an array get we try to apply the optimization, but then we have already insert the main IfElse instruction when we actually want to insert the array get and new IfElse instruction before hand.

Reply:
>  but then we have already insert the main IfElse instruction

Is the idea to remove the original IfElse instruction if possible? According to Tom's comment it can be left, then it will be removed if it doesn't survive DIE (see last sentence of #5501's description)

Reply:
No we can leave it in as per Tom's comment. I was mainly surprised to not see `v9 = if v2 then v7 else if v4 then v8`, but I do see now that we automatically simplify an `IfElse` instruction if the result type is `Type::Numeric(_)`. So this actually is expected. I want to test out this PR a bit more, going to resolve this comment though.